server/src/main/java/io/unitycatalog/server/service/TemporaryModelVersionCredentialsService.java

```diff
@@ -31,29 +31,41 @@ public TemporaryModelVersionCredentialsService(CredentialOperations credentialOp
+    public HttpResponse generateTemporaryModelVersionCredentials(
+            GenerateTemporaryModelVersionCredentials generateTemporaryModelVersionCredentials) {
 
-        long modelVersion = generateTemporaryModelVersionCredential.getVersion();
-        String catalogName = generateTemporaryModelVersionCredential.getCatalogName();
-        String schemaName = generateTemporaryModelVersionCredential.getSchemaName();
-        String modelName = generateTemporaryModelVersionCredential.getModelName();
+        long modelVersion = generateTemporaryModelVersionCredentials.getVersion();
+        String catalogName = generateTemporaryModelVersionCredentials.getCatalogName();
+        String schemaName = generateTemporaryModelVersionCredentials.getSchemaName();
+        String modelName = generateTemporaryModelVersionCredentials.getModelName();
         String fullName = RepositoryUtils.getAssetFullName(catalogName, schemaName, modelName);
 
         ModelVersionInfo modelVersionInfo = MODEL_REPOSITORY.getModelVersion(fullName, modelVersion);
-        ModelVersionOperation requestedOperation = generateTemporaryModelVersionCredential.getOperation();
-        // Must enforce that write credentials are not passed back if the model has been finalized
-        if (modelVersionInfo.getStatus() != ModelVersionStatus.PENDING_REGISTRATION && requestedOperation == ModelVersionOperation.READ_WRITE_MODEL_VERSION) {
-            throw new BaseException(ErrorCode.INVALID_ARGUMENT, "Cannot request read/write credentials on a finalized model version: " + fullName + "/" + modelVersion);
+        String storageLocation = modelVersionInfo.getStorageLocation();
+        if (storageLocation.toLowerCase().startsWith("file")) {
+            throw new BaseException(ErrorCode.INVALID_ARGUMENT, "Cannot request credentials on a model version with a file based storage location: " + fullName + "/" + modelVersion);
         }
-        return HttpResponse.ofJson(credentialOps.vendCredentialForModelVersion(modelVersionInfo, modelVersionOperationToPrivileges(generateTemporaryModelVersionCredential.getOperation())));
+        ModelVersionOperation requestedOperation = generateTemporaryModelVersionCredentials.getOperation();
+        // Must enforce that the status of the model version matches the requested credential type.
+        if (modelVersionInfo.getStatus() == ModelVersionStatus.FAILED_REGISTRATION) {
+            throw new BaseException(ErrorCode.INVALID_ARGUMENT, "Cannot request credentials on a model version that has failed registration: " + fullName + "/" + modelVersion);
+        }
+        if (modelVersionInfo.getStatus() == ModelVersionStatus.MODEL_VERSION_STATUS_UNKNOWN) {
+            throw new BaseException(ErrorCode.INVALID_ARGUMENT, "Cannot request credentials on a model version with an unknown status: " + fullName + "/" + modelVersion);
+        }
+        if ((modelVersionInfo.getStatus() != ModelVersionStatus.PENDING_REGISTRATION &&
```

Comment:
The if condition results to model version = finalized (model version != pending registration) and requestedOperation = Read_Write 
but the error message reads otherwise.